### PR TITLE
Distinguish User Data Field required for Web User or Mobile Worker

### PR DESCRIPTION
**Pull request by Jtang-1**
## Product Description

When editing a User Field, checking "Required" will also display a dropdown with the selection "Web Users", "Mobile Workers" or "Both". When
- "Web Users" is selected, that field is required only when editing a Web User or for Web User Invite.
- "Mobile Workers" is selected, that field is required only when editing a Mobile Worker and creating a Mobile Worker.
- "Both" is selected, that field is required in all cases.

Screen Shot 2024-10-22 at 2 06 51 PM

## Technical Summary

[USH-4867](https://dimagi.atlassian.net/browse/USH-4867)
[Design Spec
](https://docs.google.com/document/d/1QiTdoDQPPT5x2fEie6bvZD1qsUVhtSlaTTSzARW9jPg/edit?tab=t.0)
Commits are fairly clean, can review commit by commit.

## Feature Flag

no FF

### Safety story

Tested locally and on staging. The blast radius is limited to Custom User Data.

### Automated test coverage

Added tests
- Field property containing lists are serialized upon form submission and cleaned data contains the expected list
- Field model is serialized correctly
- Form fields that should be required are set as required upon form creation `_make_field`
- View helper function `is_field_required` to check if a field is required (taking into consideration `required_for`) 
- Filtering and getting Fields from `CustomDataFieldsDefinition` works

### QA Plan

[QA-7171
](https://dimagi.atlassian.net/browse/QA-7171)
### Migrations

- [X] The migrations in this code can be safely applied first independently of the code

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

[USH-4867]: https://dimagi.atlassian.net/browse/USH-4867?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: corehq/apps/users/views/mobile/custom_data_fields.py [hunk 1]
```diff
@@ -21,6 +21,25 @@ class UserFieldsView(CustomDataModelMixin, BaseUserSettingsView):
     _show_profiles = True
     page_title = _('Edit User Fields')
 
+    user_type = None
+    WEBUSER = "web_user"
```
[the remark this answers is not in the document]
Comment (Jtang-1) [reply corehq/apps/users/views/mobile/custom_data_fields.py]: [2f16749](https://github.com/dimagi/commcare-hq/pull/35218/commits/2f167499f229888ac517e906a9930953157f7e21)

**Comment by orangejenny** [comment]
UI thought: What if this were a dropdown of "Mobile Workers", "Web Users", "both", or "neither"? That seems simpler to think about than a checkbox plus a dropdown.

File: corehq/apps/custom_data_fields/migrations/0009_field_required_for.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+@skip_on_fresh_install
+def update_user_fields_required_for(apps, schema_editor):
+    Field = apps.get_model('custom_data_fields', 'Field')
+    CustomDataFieldsDefinition = apps.get_model('custom_data_fields', 'CustomDataFieldsDefinition')
+
+    user_fields_definitions = CustomDataFieldsDefinition.objects.filter(field_type=CUSTOM_USER_DATA_FIELD_TYPE)
+
+    for definition in user_fields_definitions:
+        Field.objects.filter(definition=definition).update(required_for=DEFAULT_REQUIRED_FOR)
```
Comment (orangejenny) [inline corehq/apps/custom_data_fields/migrations/0009_field_required_for.py:19]: Isn't this a change from current behavior? From skimming the ticket and design document, it sounds like right now fields are required for both mobile and web users. It sounds like mobile workers only is the desired default behavior, but are you certain this change won't affect existing projects in a way they don't expect?

**Comment by orangejenny** [comment]
Skimmed and left a couple of comments, didn't review closely.

File: corehq/apps/custom_data_fields/migrations/0009_field_required_for.py (hunk 2 above)
Comment (Jtang-1) [reply corehq/apps/custom_data_fields/migrations/0009_field_required_for.py:19]: Yes it is a change in current behavior and you're correct that right now, fields are required for both mobile and web users. We discussed/are discussing it in [this](https://docs.google.com/document/d/1QiTdoDQPPT5x2fEie6bvZD1qsUVhtSlaTTSzARW9jPg/edit?disco=AAABVQJrlRA) comment thread. The decision from that discussion thread was to migrate existing projects to have required fields apply to "mobile worker".

But you're right that it would be disruptive to change current behavior. So I think the better strategy would be to migrate such that required fields still apply to both web users and mobile workers. And then domains who do not want required fields applied to web users can configure it as such.

**Comment by Jtang-1** [comment]
I did consider including a "None" option in the dropdown and getting rid of the checkbox. But I'm assuming most fields are not going to be required so I think it reduces clutter to just have a checkbox and not unnecessarily display a bunch of dropdowns with text. Also, this same input template is used for other custom data fields so it would also be more consistent to use a checkbox for them all (rather than having only a dropdown here and a checkbox for the other ones)

File: corehq/apps/custom_data_fields/migrations/0009_field_required_for.py (hunk 2 above)
Comment (Jtang-1) [reply corehq/apps/custom_data_fields/migrations/0009_field_required_for.py]: Changed migration such that existing required fields continue to apply to both web users and mobile workers [fa7b8d4](https://github.com/dimagi/commcare-hq/pull/35218/commits/fa7b8d43e4223ccf382bb9803e4b707f127bea4d) [9a56a14](https://github.com/dimagi/commcare-hq/pull/35218/commits/9a56a14873ee94acb08f4a31cb932eb2fdf50c0f) Thank you for flagging this, definitely would have caused commotion if I had migrated existing fields to apply only for mobile workers.

**Comment by orangejenny** [comment]
@Jtang-1 The UI point makes sense, especially the point about most fields not being required. UI nitpick: would it be feasible to move up the dropdown so it's on the same line as the checkbox, instead of making every row twice as tall?

File: corehq/apps/custom_data_fields/tests/test_fields.py [hunk 3]
```diff
@@ -21,6 +21,48 @@ def test_validate_required(self):
         self.assertIsNone(optional_field.validate_required('sea cucumber'))
         self.assertIsNone(optional_field.validate_required(None))
 
+    def test_validate_required_for(self):
```
Comment (minhaminha) [inline corehq/apps/custom_data_fields/tests/test_fields.py:24]: This test should be broken up into 4 individual tests, so that it gets through all the scenarios without getting stuck at an earlier assert.

File: corehq/apps/custom_data_fields/tests/test_form.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+    _create_field = create_field
+
+
+class TestCustomDataEditorFormFields(SimpleTestCase, FieldsViewMixin):
+    def setUp(self):
+        self.domain = 'test-domain'
+
+    @patch('corehq.apps.custom_data_fields.edit_entity.CustomDataEditor.init_form')
+    def test_field_is_required(self, mock_init_form):
```
Comment (minhaminha) [inline corehq/apps/custom_data_fields/tests/test_form.py:42]: Same with this one - I think it should be 3 different tests (maybe put the _editor and _field vars in the setUp)

File: corehq/apps/users/migrations/__pycache__/0027_role_permission_unique.cpython-310.pyc.6183703056
Comment (minhaminha) [inline corehq/apps/users/migrations/__pycache__/0027_role_permission_unique.cpython-310.pyc.6183703056:1]: Been a while since I did a migration PR... is this file needed?

**Review by minhaminha — COMMENTED** [review]
Looks good! Just some nitpicky stuff about the tests. Also, maybe this is only a SaaS thing but I thought migrations changes should always be in their own, isolated PR, to make the revert process a bit easier in the event that needs to happen.

File: corehq/apps/users/migrations/__pycache__/0027_role_permission_unique.cpython-310.pyc.6183703056
Comment (Jtang-1) [reply corehq/apps/users/migrations/__pycache__/0027_role_permission_unique.cpython-310.pyc.6183703056:1]: thanks that was an accident, it is not needed [b7f8a1d](https://github.com/dimagi/commcare-hq/pull/35218/commits/b7f8a1d35306edd9be8846bd41cbe483606ba330)

File: corehq/apps/custom_data_fields/tests/test_form.py (hunk 4 above)
Comment (Jtang-1) [reply corehq/apps/custom_data_fields/tests/test_form.py]: good call [8dd8f94](https://github.com/dimagi/commcare-hq/pull/35218/commits/8dd8f94b8bc4a4b453438d23bbee8391b8838d61)

File: corehq/apps/custom_data_fields/tests/test_fields.py (hunk 3 above)
Comment (Jtang-1) [reply corehq/apps/custom_data_fields/tests/test_fields.py]: [8dd8f94](https://github.com/dimagi/commcare-hq/pull/35218/commits/8dd8f94b8bc4a4b453438d23bbee8391b8838d61)

File: corehq/apps/custom_data_fields/templates/custom_data_fields/partials/fields_tab.html [hunk 5]
```diff
@@ -88,17 +88,17 @@
       <input class="form-control" type="text" data-bind="value: label, attr: { 'readonly': !isEditable() }"/>
     </td>
     <td>
-      <div class="form-check mt-2">
+      <div class="form-check mt-2 d-flex align-items-center">
```
Comment (orangejenny) [inline corehq/apps/custom_data_fields/templates/custom_data_fields/partials/fields_tab.html:91]: thank you!

**Comment by Jtang-1** [comment]
@orangejenny Thanks for the suggestion, that looks better. Updated the screenshot in the description.

@minhaminha It is preferable for the migration to be in its own PR, but for simple migrations I think it's fine for the migration to be in the same PR as its associated code changes.

**Review by minhaminha — APPROVED** [review]
